src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,113 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_LENGTH = 10;
+    private static final double GROW_FACTOR = 1.5;
+    private int filledCount;
+    private T[] items;
+
+    public ArrayList() {
+        items = (T[]) new Object[INITIAL_LENGTH];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNeeded();
+        items[filledCount] = value;
+        filledCount++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index != filledCount) {
+            checkOutOfBoundsException(index);
+        }
 
+        growIfNeeded();
+
+        for (int i = filledCount; i >= index; i--) {
+            items[i] = (i == index) ? value : items[i - 1];
+        }
+        filledCount++;
     }
 
     @Override
     public void addAll(List<T> list) {
 
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkOutOfBoundsException(index);
+        return items[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkOutOfBoundsException(index);
+        items[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkOutOfBoundsException(index);
+        final T removedItem = items[index];
+        System.arraycopy(items, index + 1, items, index, filledCount - 1 - index);
+        items[filledCount - 1] = null;
+        filledCount--;
+        return removedItem;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < filledCount; i++) {
+            if (objectsAreEqual(items[i], element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Can't find element " + element + " in the list");
     }
 
     @Override
     public int size() {
-        return 0;
+        return filledCount;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (size() == 0);
+    }
+
+    private boolean isFull() {
+        return (filledCount == items.length);
+    }
```

Comment:
redundant parentheses 
logic is trivial and only used once - put it where this method is called instead